### ISSUE TITLE
GRPO with tool calling

### ISSUE THREAD

**Issue by accupham** [huggingface/trl#2712]
### Method description

Would it be possible to implement an RL environment that does multi-turn tool calling in the GRPO RL training loop? Right now it seems to be a one-shot inference before passing it to the custom reward function. I'd like to have a multi-turn interaction via tool calling step before passing the final result to the reward function.

Online tool calling would enable RL over a simulation with feedback from the environment. Seems vLLM supports all manner of tool calling. Could this be added as part of GRPO?

@qgallouedec   

### Open source status

- [ ] The method implementation is available
- [ ] The model weights are available
- [ ] The training datasets are available

**Comment by August-murr** [comment]
#2704

**Comment by percent4** [comment]
+1. I want to use tool, such as Python code, to solve math problem with GRPO, but there is no support for this. Hope the trl can support GRPO with Tool Use.

**Comment by big-nico** [comment]
+1. Other RL training libraries like VeRL support this. Would be great to be able to use TRL for this. Right now the lack of multi-step agentic behavior really limits the agentic work you can do.

**Comment by pietergbosma** [comment]
I created this with unsloth, check out [(https://discord.com/channels/1179035537009545276/1343968352875319336)](url)

**Comment by cfpark00** [comment]
@big-nico  Verl supports this? can you show me where? Thank you!

**Comment by willccbb** [comment]
TRL-based project which supports tool calling + code execution: https://github.com/willccbb/verifiers

**Comment by qgallouedec** [comment]
Closing this issue as this has been supported for quite some time now, see https://huggingface.co/docs/trl/v1.4.0/en/grpo_trainer#agent-training
